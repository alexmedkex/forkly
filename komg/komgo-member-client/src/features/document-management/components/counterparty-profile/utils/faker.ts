import { CounterpartyProfile, RiskLevel } from '../../../../counterparties/store/types'

const defaultProfile: CounterpartyProfile = {
  id: 'anon-profile-id',
  counterpartyId: 'anon-counterparty-id',
  renewalDate: new Date('2019-07-16').toISOString(),
  riskLevel: RiskLevel.unspecified,
  managedById: 'anon-user-id'
}

export const fakeProfile = (partial?: Partial<CounterpartyProfile>) => {
  return partial ? { ...defaultProfile, ...partial } : defaultProfile
}

export const mockDate = () => {
  const RealDate = Date
  const _GLOBAL: any = global // we love typescript!

  return {
    freeze: (isoDate: string | number | Date) => {
      _GLOBAL.Date = class extends RealDate {
        constructor(...args: any) {
          super()
          return new RealDate(isoDate)
        }
      }
    },
    restore: () => {
      global.Date = RealDate
    }
  }
}
