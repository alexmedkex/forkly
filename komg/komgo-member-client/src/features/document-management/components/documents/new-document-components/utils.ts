import { HasId, HasName } from '../../../store/types'

export const hasIdAndNameToOption = (collection: Array<HasId & HasName>): Array<{ key: string }> => {
  const originalCollection = collection.map(el => {
    return { key: el.id, text: el.name, value: el.id }
  })
  return originalCollection.reduce((unique: Array<{ key: string }>, o: { key: string }) => {
    if (!unique.some((obj: { key: string }) => obj.key === o.key)) {
      unique.push(o)
    }
    return unique
  }, [])
}
