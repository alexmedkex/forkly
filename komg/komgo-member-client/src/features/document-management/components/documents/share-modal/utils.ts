import { ICompanyTableItemWithSharedDocs, ICounterpartyWithDocuments } from './SelectedCounterparties'
import { FlagNameValues } from 'semantic-ui-react'
import * as i18nIsoCountries from 'i18n-iso-countries'

export function pluralize(word: string, count: number, singleSuffix: string, pluralSuffix: string) {
  if (count > 1) {
    return `${word}${pluralSuffix}`
  }

  return `${word}${singleSuffix}`
}

export const buildCounterpartyWithDocumentsPickerItems = (
  counterpartiesWithDocs: ICounterpartyWithDocuments[]
): ICompanyTableItemWithSharedDocs[] => {
  i18nIsoCountries.registerLocale(require('i18n-iso-countries/langs/en.json'))
  return counterpartiesWithDocs.map(item => ({
    name: item.counterparty.x500Name.CN,
    countryName: i18nIsoCountries.getName(item.counterparty.x500Name.C.trim(), 'en'),
    country: item.counterparty.x500Name.C.toLocaleLowerCase().trim() as FlagNameValues,
    location: item.counterparty.x500Name.L,
    id: item.counterparty.staticId,
    sharedDocuments: item.documents
  }))
}
