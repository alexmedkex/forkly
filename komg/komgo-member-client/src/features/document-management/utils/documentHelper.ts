import { Document } from '../store/types'

export const isSharedDocument = (doc: Document): boolean => {
  return doc.sharedBy !== 'none'
}

/**
 * It checks if a document have been already reviewed or it returns
 * false if it is pending to be reviewed yet.
 */
export const isReviewed = (doc: Document): boolean => {
  if (isSharedDocument(doc)) {
    return doc.sharedInfo && doc.sharedInfo.status && doc.sharedInfo.status !== 'pending'
  }
}

export const isReviewCompleted = (doc: Document): boolean => {
  if (doc.sharedInfo) {
    return doc.sharedInfo.completedReview
  }
}
