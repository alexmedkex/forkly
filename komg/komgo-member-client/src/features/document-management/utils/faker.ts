import { Document } from '../store/types'
import { mockProduct } from '../store/products/mock-data'
import { mockCategories } from '../store/categories/mock-data'
import { mockDocumentTypes } from '../store/document-types/mock-data'
import { BottomSheetStatus } from '../../bottom-sheet/store/types'
import { DropdownOption } from '../components/documents/my-documents/DocumentListDropdownOptions'

export const MOCK_COMPANY_ID = 'company-id'

export const anonDocument: Document = {
  id: '-1',
  documentId: 'anonDocumentId',
  name: 'anon document',
  product: mockProduct,
  category: mockCategories[2],
  type: mockDocumentTypes[2],
  owner: {
    firstName: 'Owner',
    lastName: 'Owner',
    companyId: MOCK_COMPANY_ID
  },
  hash: 'hash',
  receivedDate: new Date('2000-12-31'),
  registrationDate: new Date('2000-12-31'),
  metadata: [],
  content: undefined,
  sharedWith: [],
  sharedBy: '',
  state: BottomSheetStatus.PENDING
}

export const fakeDocument = (partialDocument?: Partial<Document>): Document => {
  const partial = partialDocument || {}
  return { ...anonDocument, ...partial }
}

const defaultDropdownOption: DropdownOption = {
  key: 'anon',
  text: '',
  value: 'anon',
  content: 'Anon',
  onClick: () => void 0
}

export const fakeDropdownOption = (partial: Partial<DropdownOption>): DropdownOption => {
  return { ...defaultDropdownOption, ...partial }
}
