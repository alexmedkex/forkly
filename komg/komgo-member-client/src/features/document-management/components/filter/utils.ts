import { Category, DocumentType } from '../../index'
import { sortCategories, sortDocumentTypes } from '../../utils/sortingHelper'
import { categoryToColor } from '../documents/document-library/categoryToColor'

export const getDocumentTypeFilterOptions = (categories: Category[], types: DocumentType[]) => {
  return sortCategories(categories).map(c => ({
    name: c.name,
    value: c.id,
    display: {
      color: categoryToColor[c.id]
    },
    items: sortDocumentTypes(types)
      .filter(t => t.category.id === c.id)
      .map(t => ({
        name: t.name,
        value: t.id
      }))
  }))
}
