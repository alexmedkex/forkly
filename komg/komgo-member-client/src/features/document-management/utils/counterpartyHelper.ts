import { Counterparty } from '../../counterparties/store/types'
import { findCounterpartyByStatic } from '../../letter-of-credit-legacy/utils/selectors'

export const counterpartyName = (counterparty: Counterparty): string => {
  return counterparty.x500Name.O
}

export const fetchCounterpartyName = (counterparties: Counterparty[], companyId: string): string => {
  const counterParty = findCounterpartyByStatic(counterparties, companyId)
  return counterParty ? counterParty.x500Name.CN : ''
}
