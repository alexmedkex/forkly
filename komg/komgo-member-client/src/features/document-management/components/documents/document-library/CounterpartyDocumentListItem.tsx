import * as React from 'react'
import styled from 'styled-components'
import { Button, Icon } from 'semantic-ui-react'

import { compose } from 'redux'
import { withRouter } from 'react-router'
import { RouteComponentProps } from 'react-router'
import { Document, SharedInfo } from '../../../store'
import { toMegabytes, truncate } from '../../../../../utils/casings'
import { displayDate } from '../../../../../utils/date'
import { DropdownOption } from '../my-documents/DocumentListDropdownOptions'
import * as DocumentHelper from '../../../utils/documentHelper'
import StatusPill from '../../../../review-documents/components/StatusPill'
import CommentTooltip from '../../../../review-documents/components/CommentTooltip'
import { ReviewStatus } from '../../../../review-documents/store/types'

import { ListItemField, Field, ListItemLabel, ListItemValue } from './ListItemField'

export interface Props extends RouteComponentProps<any> {
  document: Document
  viewDocumentOption: DropdownOption
  downloadDocumentOption: DropdownOption
  highlighted: boolean
  getUserNameFromDocumentUploaderId(idUser: string): string
  renderEllipsisMenu(doc: Document): React.ReactNode
}

interface ReviewProps {
  isReviewCompleted: boolean
}

const CounterpartyDocumentListItem: React.SFC<Props> = (props: Props) => {
  const doc = props.document
  const [docName, ext] = doc.name.split('.')
  const size = getDocumentSize(doc.content as any)
  const docIsReviewCompleted = DocumentHelper.isReviewCompleted(doc)

  return (
    <ListItemGrid isReviewCompleted={docIsReviewCompleted}>
      <FieldGrid isReviewCompleted={docIsReviewCompleted} data-test-id="field-grid">
        <ListItemField label="Type" value={doc.type.name} dataTestId="field-type" />
        <ListItemField label="Format" value={ext.toUpperCase()} dataTestId="field-format" />
        <ListItemField
          label="Received on"
          value={displayDate(doc.receivedDate, 'DD MMM YYYY')}
          dataTestId="field-uploaded"
        />
        <ListItemField label="Name" value={docName} dataTestId="field-name" />
        <ListItemField label="Size" value={size} dataTestId="field-size" />
        <ReviewStatusField data-test-id="field-status">
          <ListItemLabel>Review Status</ListItemLabel>
          <ListItemValue>{displayStatus(props)}</ListItemValue>
        </ReviewStatusField>
      </FieldGrid>
      {renderActionButtons(props)}
    </ListItemGrid>
  )
}

const renderActionButtons = (props: Props): React.ReactNode => {
  const doc: Document = props.document
  if (DocumentHelper.isSharedDocument(doc)) {
    if (DocumentHelper.isReviewCompleted(doc)) {
      // In case it is a shared document and its review is not completed yet
      return (
        <>
          <div data-test-id="view-button">{renderOptionButton(props.viewDocumentOption)}</div>
          <div data-test-id="download-button">{renderOptionButton(props.downloadDocumentOption)}</div>
        </>
      )
    } else {
      /* In case the review process of this document has not been completed yet we can 
      still review it as many times as we want */
      return (
        <>
          <Button data-test-id="review-button" onClick={() => onReviewClick(props)} primary={true}>
            Review
          </Button>
          {props.renderEllipsisMenu(doc)}
        </>
      )
    }
  }
}

const displayStatus = (props: Props) => {
  const doc: Document = props.document
  if (DocumentHelper.isSharedDocument(doc)) {
    if (DocumentHelper.isReviewed(doc)) {
      return renderStatusReviewed(props)
    } else {
      return 'Awaiting'
    }
  }
}

const onReviewClick = (props: Props) => {
  const doc: Document = props.document
  props.history.push({
    pathname: '/evaluation',
    state: {
      documents: [{ document: doc, status: doc.sharedInfo.status, note: doc.sharedInfo.note }],
      documentId: doc.id,
      sendDocumentsRequestId: doc.sharedInfo.receivedDocumentsId,
      redirectBackUrl: undefined
    }
  })
}

const renderStatusReviewed = (props: Props) => {
  const shared = props.document.sharedInfo
  return (
    <>
      <StatusPill status={shared.status} />
      <ListItemLabel>BY</ListItemLabel>
      <ListItemValue data-test-id="field-reviewed-by">
        {props.getUserNameFromDocumentUploaderId(shared.reviewerId)}
      </ListItemValue>
      {shared.status === ReviewStatus.REJECTED && shared.note !== '' ? (
        <ListItemLabel data-test-id="field-info-bubble">
          <CommentTooltip comment={shared.note} icon={<Icon size="large" name="comment alternate" />} />
        </ListItemLabel>
      ) : (
        ''
      )}
    </>
  )
}

const renderOptionButton = (option: DropdownOption) => {
  const enabled = option && !option.disabled
  return (
    <Button disabled={!enabled} onClick={option.onClick}>
      {option.value}
    </Button>
  )
}

const getDocumentSize = (content: { size: string }) => (content && content.size ? toMegabytes(content.size) : 'Unknown')

const ListItemGrid = styled.div`
  height: 72px;
  border: 1px solid #e8eef3;
  border-radius: 4px;
  background-color: #ffffff;
  display: grid;
  grid-template-columns: ${(props: ReviewProps) =>
    props.isReviewCompleted
      ? 'auto [fields] 70px [view-button] 110px [download-button]'
      : 'auto [fields] 80px [review-button] 50px [ellipses-button]'};
  align-items: center;
`

const FieldGrid = styled.div`
  display: grid;
  grid-template-columns: ${(props: ReviewProps) =>
    props.isReviewCompleted ? '0.65fr 0.65fr 1fr' : '0.58fr 0.58fr 1fr'};
  grid-template-rows: 28px;
  grid-column-gap: 2rem;
`

const ReviewStatusField = styled(Field)`
  grid-template-columns: repeat(4, minmax(2em, max-content));
`

export default compose<any>(withRouter)(CounterpartyDocumentListItem)
