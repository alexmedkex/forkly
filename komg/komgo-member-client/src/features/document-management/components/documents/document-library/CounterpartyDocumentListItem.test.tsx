import * as React from 'react'
import * as renderer from 'react-test-renderer'
import CounterpartyDocumentListItem from './CounterpartyDocumentListItem'
import { Props } from './CounterpartyDocumentListItem'
import { createMemoryHistory } from 'history'
import { fakeDocument, fakeDropdownOption } from '../../../utils/faker'
import { MemoryRouter as Router } from 'react-router-dom'

describe('CounterpartyDocumentListItem', () => {
  const mockDocument = fakeDocument({ name: 'anon.ext', uploadInfo: { uploaderUserId: '-1' } })
  const mockProps: Props = {
    history: createMemoryHistory(),
    staticContext: undefined,
    location: {
      pathname: '',
      search: '',
      state: '',
      hash: ''
    },
    match: undefined,
    document: mockDocument,
    highlighted: false,
    viewDocumentOption: fakeDropdownOption({ key: 'view', value: 'anon', onClick: jest.fn() }),
    downloadDocumentOption: fakeDropdownOption({ onClick: jest.fn(), key: 'download', value: 'anon' }),
    getUserNameFromDocumentUploaderId: jest.fn((idUser: string) => 'no-one'),
    renderEllipsisMenu: doc => <div>ellipsis</div>
  }
  it('renders', () => {
    expect(
      renderer
        .create(
          <Router>
            <CounterpartyDocumentListItem {...mockProps} />
          </Router>
        )
        .toJSON()
    ).toMatchSnapshot()
  })
})
