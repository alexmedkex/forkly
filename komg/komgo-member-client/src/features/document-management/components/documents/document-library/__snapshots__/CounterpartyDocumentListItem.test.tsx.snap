// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterpartyDocumentListItem renders 1`] = `
.c2 {
  display: grid;
  grid-template-columns: repeat(2,minmax(4.2em,max-content));
}

.c3 {
  margin: 0 8px;
  color: #7f95aa;
  font-size: 11px;
  font-weight: 600;
  text-transform: uppercase;
  line-height: 21px;
}

.c4 {
  color: #1c2936;
  font-size: 14px;
  line-height: 21px;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c0 {
  height: 72px;
  border: 1px solid #e8eef3;
  border-radius: 4px;
  background-color: #ffffff;
  display: grid;
  grid-template-columns: auto [fields] 80px [review-button] 50px [ellipses-button];
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: grid;
  grid-template-columns: 0.58fr 0.58fr 1fr;
  grid-template-rows: 28px;
  grid-column-gap: 2rem;
}

.c5 {
  grid-template-columns: repeat(4,minmax(2em,max-content));
}

<div
  className="c0"
>
  <div
    className="c1"
    data-test-id="field-grid"
  >
    <div
      className="c2"
      data-test-id="field-type"
    >
      <label
        className="c3"
      >
        Type
      </label>
      <span
        className="c4"
      >
        Passports of UBOs
      </span>
    </div>
    <div
      className="c2"
      data-test-id="field-format"
    >
      <label
        className="c3"
      >
        Format
      </label>
      <span
        className="c4"
      >
        EXT
      </span>
    </div>
    <div
      className="c2"
      data-test-id="field-uploaded"
    >
      <label
        className="c3"
      >
        Received on
      </label>
      <span
        className="c4"
      >
        31 Dec 2000
      </span>
    </div>
    <div
      className="c2"
      data-test-id="field-name"
    >
      <label
        className="c3"
      >
        Name
      </label>
      <span
        className="c4"
      >
        anon
      </span>
    </div>
    <div
      className="c2"
      data-test-id="field-size"
    >
      <label
        className="c3"
      >
        Size
      </label>
      <span
        className="c4"
      >
        Unknown
      </span>
    </div>
    <div
      className="c5 c2"
      data-test-id="field-status"
    >
      <label
        className="c3"
      >
        Review Status
      </label>
      <span
        className="c4"
      >
        Awaiting
      </span>
    </div>
  </div>
  <button
    className="ui primary button"
    data-test-id="review-button"
    onClick={[Function]}
    role="button"
  >
    Review
  </button>
  <div>
    ellipsis
  </div>
</div>
`;
